src/Sylius/Component/Inventory/Operator/BackordersHandler.php

```diff
@@ -78,16 +86,19 @@ public function fillBackorders(StockableInterface $stockable)
             return;
         }
 
-        $units = $this->repository->findBy([
-            'stockable' => $stockable,
-            'inventoryState' => InventoryUnitInterface::STATE_BACKORDERED,
-        ], ['createdAt' => 'ASC']);
+        $units = $this->repository->findByStockableAndInventoryState(
+            $stockable,
+            InventoryUnitInterface::STATE_BACKORDERED,
+            $onHand
+        );
 
         foreach ($units as $unit) {
-            $unit->setInventoryState(InventoryUnitInterface::STATE_SOLD);
+            $sm = $this->factory->get($unit, InventoryUnitTransitions::GRAPH);
 
-            if (--$onHand === 0) {
-                break;
+            if ($sm->apply(InventoryUnitTransitions::SYLIUS_HOLD)) {
```

Comment:
IMO, I think that you want an exception to be thrown, like as it is now. When somehow the transition can't be applied it is better that the coder knows what is up. For example when he implements his own `findByStockableAndInventoryState` and makes a mistake there by not always returning backordered item units.